design/dhcpless_ipam.md

```diff
@@ -0,0 +1,472 @@
+do not define a specific solution as part of this proposal.
+
+### Test Plan
+
+* TODO - test coverage, is there any existing coverage for the IPAM flow we can extend?
+
+### Upgrade / Downgrade Strategy
+
+This will be a net-new API and controller so there should be no upgrade impact.
+
+Downgrade of environments leveraging the new API will not be possible, but given provisioning is not currently
+possible via this specific flow that seems unlikely to be an issue.
+
+### Version Skew Strategy
+
+N/A
+
+## Drawbacks
+
+### Not all disk formats are supported
+
+This approach will not work in some configurations, and specifically it won't work with `format: live-iso`
+since in that case Ironic doesn't build the virtualmedia image, so it can't inject any `network_data`
+
+Additionally Ironic may not support injecting `network_data` in every configuration, when the IPA image is
+supplied as kernel and initramfs combo then Ironic can convert it to an ISO and inject the node `network_data`
+but in the case where a pre-built ISO ramdisk ISO is provided.
+
+In practice this probably isn't a huge issue, since if the preprovisioning image workflow is used to generate
+a customized ISO, it can embed the configuration instead of Ironic to achieve the same result.
+
+### Network configuration failure can't be reflected in status
+
+In the case of failure to configure the pre-provisioning interfaces, the provisioning agent can't reply to
+indicate status/failure so the node inspection will just hang without an explicit/useful error message.
+
+I don't think there is an obvious solution to this, other than potentially considering some kind of validation
+but that's outside the scope of this proposal.
+
+### OpenStack networkData format
+
+This solution builds on our existing CAPM3 networkData templating, and as such is is limited to configuration
+tools that can interpret OpenStack network,user or metadata formats.
+
+In future it would be interesting to explore alternative options, for example using [nmstate](https://nmstate.io/)
+as the DSL where the nmstate CLI tool could interpret the data independent of specific firstboot tool choice.
+
+This proposal has tries to account for such a future direction, but such implementation would imply pluggable
+templating which is outside the scope of this change, and may be considered via a future proposal.
+
+## Alternatives
+
+### BMO change alternatives
+
+There are several alternative approaches to solve the problem of preventing BMO starting inspection immediately on BMH registration:
+
+* Add a new BareMetalHost API `PreprovisioningNetworkDataRequired` which defaults to false, but when set to true will describe that the host cannot move from Registering -> Inspecting until `preprovisioningNetworkDataName` has been set.
+* Create the BMH with a preprovisioningNetworkDataName pointing to an empty Secret. BMO refuses to start inspecting until the Secret contains some data.
```

Comment:
The principle that necessary data can be specified as non existing resource and that the BMH will just wait and will not err until the data is provided is nice. Even with the current approach, this could be used by template custom resource that can generate the data but without the power to modify the BMH fields. And the principle can be generalized to other data (userData, networkData).